internal/stores/contracts_test.go

```diff
@@ -23,6 +26,13 @@ func TestSQLContractStore(t *testing.T) {
 	uc, _ := types.GenerateDeterministicMultisig(1, 2, "salt")
 	uc.Timelock = 192837
 
+	// Create a host for the contract.
+	hk := consensus.GeneratePrivateKey().PublicKey()
+	err = cs.RecordInteraction(hk, hostdb.Interaction{Timestamp: time.Now()})
```

Comment:
maybe we should add `AddHost` to the `HostDB` interface, I know this 'll do a `FirstOrCreate` but it's better to be explicit

Reply:
And have `RecordInteraction` and `RecordAnnouncement` fail without the host being added first?

Reply:
I think I prefer the existing behavior.

"Explicit" is a slippery word. You could argue: "in order to do X, Y has to happen first, so requiring the caller to do Y is more explicit." The problem is: where do you stop? For example, in order to add a host to the DB, the DB has to allocate space for it; should the DB have an `Allocate()` function that must be called before each insertion? And `Allocate` of course has its own internal sequence of steps that must be performed; should all of those be inlined at the callsite too?

So we see that this is really a question of *responsibility*. All of these actions need to happen at *some* point, but who should be responsible for performing them? It depends on what sort of API you want: how much power you want to expose, or how idiot-proof you want it to be, etc. In this particular case, I think a distinct "add host" step is not relevant to our desired API. What we care about is associating an interaction or announcement with a host key -- that's it. The caller doesn't care whether an entry for the host already exists or not! And consequently, it should not be their responsibility.

Reply:
> And have `RecordInteraction` and `RecordAnnouncement` fail without the host being added first?

no

Reply:
I think my comment was misinterpreted - also maybe I should have added `nit` - I was proposing to add an `AddHost` function because a) it seems like a pretty basic method to have in the interface, even though it would only be used in testing for the time being, and b) I didn't like the implicitness of inserting a host in the db through an announcement.

You are essentially assuming the reader knows about the behaviour in `RecordInteraction`. 

```go
// create host for the contract
hk := consensus.GeneratePrivateKey().PublicKey()
err = cs.RecordInteraction(hk, hostdb.Interaction{Timestamp: time.Now()})
```
vs 
```go
// create host for the contract
hk := consensus.GeneratePrivateKey().PublicKey()
err = cs.AddHost(hk)
```

not worth discussing though imo it was just a remark, reading "create host" followed by "record interaction" triggered me a bit is all

Reply:
ah, I see.

... do we even *need* to call `RecordInteraction` here? Why do we need to insert a host at all? Will `AddContract` fail otherwise? (I certainly hope not!)

Reply:
We have that right now because we have a foreign key relation between contracts and hosts. Every contract in the contract table is linked to a host in the host table. You can't have contracts that link to non-existent hosts.

It is kinda nice because you never end up with inconsistent state. e.g. if you do an ip-range check you are guaranteed that you can get a net address from a contract because the host needs to exist in the db.
Vice versa you won't be able to accidentally delete a host that you have a contract with.

Of course if we don't consider the binding between contract <-> host that strong and want to handle the edge cases during contract maintenance manually, we can break that relation.